### DISCUSSION TITLE
How to handle a failed API request that is not specific to a field?

### DISCUSSION THREAD

**Discussion by eduards** [react-hook-form/react-hook-form#8365] [Q&A]
If the API request to submit the form data fails with an error that is not specific to a single form field (For example, it's a generic 500 Internal Server or 403 Forbidden or something similar), I'd like to show an error message at the bottom of the form (eg "Form submission failed due to an unexpected error. Please try again later."). How would I do this with react-hooks-form?

When I re-throw the error in `handleSubmit`, `formState.isSubmitSuccessful` would be correctly set to `false`, but I don't have access to the error message in a way I would have is if I would use `setError('fieldName', 'my custom error message)`. Furthermore, the error would bubble up to the global scope and result in an error message on the console. That's not necessarily what I want in all cases.

From [docs](https://react-hook-form.com/api/useform/seterror) for `setError` mention the following:

> Can be useful in the handleSubmit method when you want to give error feedback to a user after async validation. (ex: API returns validation errors). I can achieve what I want with the following code:

```
  const { register, handleSubmit, setError, formState: { errors } } = useForm<{ email: string }>()

  const onSubmit = handleSubmit(async (formData) => {
    try {
      await window.fetch('https://api.example.com')
    } catch (err) {
      setError('apiError', { message: 'custom error message that I want to show to the user' })
    }
  })

  ...
```
This approach is described in this article here, too: https://jasonwatmore.com/post/2021/09/13/react-hook-form-display-custom-error-message-returned-from-api-request. With this approach, `formState.isSubmitSuccessful` stays false and I can get the error message from `formState.errors.apiError`, similar to how I would retrieve the error message for specific form fields. 

However, the above doesn't fly with Typescript as `apiError` is not a form field. I could add `apiError` to the form fields type (`useForm<{ email: string, apiError?: string }>()`) but this seems wrong as `apiError` is not a form field. I'd use it only to store the error message.

Is this a shortcoming of the Typescript types or is this approach not recommended? What would be the recommended approach?

**Comment by leapful** [comment]
@eduards : Could you please share more detail about the reason why you want to handle API error with react-hook-form ? Usually, API error is not related to the form and should be handled in separated state. For example:

```
const [apiError, setApiError] = useState();

const onSubmit = handleSubmit(() => {
  try {
  } catch () {
     setApiError('Custom error message');
  }
});
```
